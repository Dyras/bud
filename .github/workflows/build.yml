name: build

on:
  push:
    paths:
      - packages/**/*/package.json
      - packages/**/*/src/*
      - packages/**/*/src/**/*
      - yarn.lock
    pull_request:
      branches:
        - stable
  workflow_dispatch:

jobs:
  build:
    strategy:
      matrix:
        platform:
          - ubuntu-latest
        node:
          - 15.11.0

    name: test node@${{matrix.node}}/${{matrix.platform}}
    runs-on: ${{matrix.platform}}
    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0

      - name: Setup node
        uses: actions/setup-node@v1
        with:
          node-version: ${{matrix.node}}

      - name: Tap cache
        id: yarn-cache-dir-path
        run: echo "::set-output name=dir::$(yarn config get cacheFolder)"

      - uses: actions/cache@v2
        id: yarn-cache
        with:
          path: ${{ steps.yarn-cache-dir-path.outputs.dir }}
          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-yarn-

      - name: Install dependenies
        run: yarn --immutable

      - name: Build project
        run: yarn build:ci
        env:
          NODE_OPTIONS: '--max-old-space-size=2048'
