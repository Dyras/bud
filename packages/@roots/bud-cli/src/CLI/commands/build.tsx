import {join} from 'path'
import {Error} from '@roots/bud-dashboard'

export const cmd = CLI => ({
  command: 'build',
  describe: 'Compile assets',
  builder: ({example}) =>
    example('Build', `${CLI.command} build --mode production`)
      .hide('help')
      .hide('version'),

  handler: ({config}) => {
    try {
      require(join(
        process.cwd(),
        (config ??
          (`${CLI.command}.config.js` as unknown)) as string,
      ))
    } catch (error) {
      Error(error.toString(), `Error`)
    }
  },
})
