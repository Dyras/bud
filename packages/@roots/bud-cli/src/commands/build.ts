import {Error} from '@roots/bud-dashboard'
import Command from '../Command'
import * as source from '../source'

/**
 * Publish
 */
export default class Build extends Command {
  public name = `build`

  public signature: Command['signature'] = '<mode>'

  public description: Command['description'] =
    'Compile assets and/or initialize development server'

  public arguments: Command['arguments'] = {
    mode: '"development" or "production"',
  }

  public options: Command['options'] = {
    src: {
      flags: '--src [src]',
      description: 'Directory to compile from',
      default: 'src',
      optional: true,
    },
    dist: {
      flags: '--dist [dist]',
      description: 'Directory to compile to',
      default: 'dist',
      optional: true,
    },
    storage: {
      flags: '--storage [storage]',
      description:
        'Directory to store build artifacts, caches, logs, etc.',
      default: '.bud',
      optional: true,
    },
    modules: {
      flags: '--modules [modules]',
      description: 'node_modules directory',
      default: 'node_modules',
      optional: true,
    },
    ci: {
      flags: '--ci [ci]',
      description: 'Run without using the Bud dashboard',
      default: false,
      optional: true,
      choices: ['true', 'false'],
    },
    debug: {
      flags: '--debug [debug]',
      description:
        'Enable debug mode. Enables logger and generates a webpack config artifact (saved to `storage`)',
      default: false,
      optional: true,
      choices: ['true', 'false'],
    },
    clean: {
      flags: '--clean [clean]',
      description:
        'Clean stale assets from dist directory during compilation',
      default: true,

      optional: true,
      choices: ['true', 'false'],
    },
    devtool: {
      flags: '--devtool [devtool]',
      description: 'Specify a sourcemap implementation',
      default: false,
      optional: true,
    },
    log: {
      flags: '--log [log]',
      description: 'Present logger outputin the terminal',
      default: false,
      optional: true,
      choices: ['true', 'false'],
    },
    manifest: {
      flags: '--manifest [manifest]',
      description: 'Generate a manifest.json file',
      default: true,

      optional: true,
      choices: ['true', 'false'],
    },
    minify: {
      flags: '--minify [minify]',
      description: 'Minify compiled assets',
      default: false,
      optional: true,
      choices: ['true', 'false'],
    },
    runtime: {
      flags: '--runtime [runtime]',
      description: 'Enable code splitting',
      default: false,
      optional: true,
      choices: ['true', 'false'],
    },
    vendor: {
      flags: '--vendor [vendor]',
      description: 'Separate application and vendor code',
      default: false,
      optional: true,
      choices: ['true', 'false'],
    },
  }

  public action({mode}) {
    try {
      source.preflight()
      source.isStatic() ? source.json() : source.api()
    } catch (error) {
      Error(error.toString(), `Error`)
    }
  }
}
