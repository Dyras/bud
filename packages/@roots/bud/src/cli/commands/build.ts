import {flags} from '@oclif/command'

import {Bud} from '../../Bud'
import {Command} from '../Command'
import {Runner} from '../Runner'

export default class Build extends Command {
  public static description = 'compile assets'

  public static examples = [
    `$ bud build production`,
    `$ bud build development`,
    `$ bud build --cache`,
  ]

  public static args = [{name: 'mode'}]

  public static flags = {
    help: flags.help({char: 'h'}),

    cache: flags.boolean({
      char: 'c',
      allowNo: true,
      default: true,
      description: 'cache compiler references to disk',
    }),

    ci: flags.boolean({
      description: 'non raw mode tty interoperable output',
    }),

    discover: flags.boolean({
      char: 'd',
      description: 'automatically utilize installed extensions',
    }),

    html: flags.boolean({
      char: 'd',
      description: 'generate an html template',
    }),

    install: flags.boolean({
      char: 'i',
      description: 'ensure peer dependencies are installed',
    }),

    log: flags.boolean({
      char: 'l',
      description: 'log to console',
    }),

    hash: flags.boolean({
      char: 'h',
      allowNo: true,
      description: 'hash compiled filenames',
    }),

    manifest: flags.boolean({
      allowNo: true,
      default: true,
      description: 'produce a manifest',
    }),

    minimize: flags.boolean({
      char: 'm',
      allowNo: true,
      description: 'minimize file size of compiled assets',
    }),

    target: flags.string({
      char: 't',
      description: 'limit compilation to this compiler',
      multiple: true,
      default: [],
    }),
  }

  public app: Bud

  public async run() {
    const runner = new Runner(this.parse(Build))

    this.app = await runner.make()

    this.app.hooks.on('done', [this.notifier.notify])

    this.app.run()
  }
}
