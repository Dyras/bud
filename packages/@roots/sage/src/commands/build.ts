import {Error} from '@roots/bud-dashboard'
import * as source from './source'
import {commands} from '@roots/bud-cli'

export const command = {
  ...commands.build,
  handler: () => {
    try {
      source.preflight()
      source.isStatic() ? source.json() : source.api()
    } catch (error) {
      Error(error.toString(), `Error`)
    }
  },
}
