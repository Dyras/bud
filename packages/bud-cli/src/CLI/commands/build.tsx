import {yargs} from '@roots/bud-support'
import {join} from 'path'
import {Error} from '../../'

const cwd = process.cwd()

export const aliases: yargs.CommandModule['aliases'] = 'build'
export const describe: yargs.CommandModule['describe'] =
  'Compile source assets.'

export const builder: yargs.CommandModule['builder'] = ({
  example,
}) =>
  example('Build', 'bud build --mode production')
    .hide('help')
    .hide('version')

export const handler: yargs.CommandModule['handler'] = (args: {
  [config: string]: unknown
}): void => {
  const cfgName = ((args.config ??
    'bud.config.js') as unknown) as string

  const cfgPath = join(cwd, cfgName)

  try {
    require(cfgPath)
  } catch (error) {
    Error(error.toString())
    process.exit()
  }
}
