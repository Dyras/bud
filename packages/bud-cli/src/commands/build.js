"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handler = exports.builder = exports.describe = exports.aliases = void 0;
const path_1 = require("path");
const __1 = require("../");
const cwd = process.cwd();
exports.aliases = 'build';
exports.describe = 'Compile source assets.';
const builder = ({ option, }) => option('config', {
    describe: 'Specify a custom configuration file',
    default: 'bud.config.js',
})
    .option('env', {
    describe: 'Build mode',
    choices: ['development', 'production', 'none'],
    default: 'none',
})
    .option('minify', {
    describe: 'Minify assets',
    type: 'boolean',
    default: false,
})
    .option('gzip', {
    describe: 'Gzip static assets',
    type: 'boolean',
    default: false,
})
    .option('hash', {
    describe: 'Hash asset filenames',
    type: 'boolean',
    default: false,
})
    .option('brotli', {
    describe: 'Apply brotli compression to static assets',
    type: 'boolean',
    default: false,
})
    .option('runtime', {
    describe: 'Generate runtime chunk',
    type: 'boolean',
    default: false,
})
    .option('vendor', {
    describe: 'Generate vendor chunk',
    type: 'boolean',
    default: false,
})
    .option('src', {
    describe: 'Override src directory',
    type: 'string',
})
    .option('dist', {
    describe: 'Override dist directory',
    type: 'string',
})
    .option('port', {
    describe: 'Dev port',
    type: 'number',
    default: 3000,
})
    .option('host', {
    describe: 'Dev host',
    type: 'string',
    default: 'localhost',
})
    .option('proxy', {
    describe: 'Hostname to proxy',
    type: 'string',
})
    .option('html', {
    describe: 'Use an HTML5 boilerplate',
    type: 'boolean',
})
    .option('log', {
    describe: 'Specify filepath to log to',
    type: 'string',
})
    .example('Dev', 'bud build --env development')
    .example('Build', 'bud build --env production --runtime --vendor --minify --gzip')
    .hide('help')
    .hide('version')
    .showHelpOnFail(true);
exports.builder = builder;
/**
 * handler: bud build
 */
const handler = (args) => {
    var _a;
    const cfg = (_a = args.config) !== null && _a !== void 0 ? _a : 'bud.config.js';
    const cfgPath = path_1.join(cwd, cfg);
    try {
        require(cfgPath);
    }
    catch (error) {
        __1.Error(`To generate a new config file try \`bud publish bud.config.js\`\nIf you are using a config file named something other than bud.config.js  you must specify its name with the --config option.\n\n${error}`, `There was a problem loading ${cfgPath}`);
    }
};
exports.handler = handler;
//# sourceMappingURL=build.js.map