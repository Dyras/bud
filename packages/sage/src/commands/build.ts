import {Error} from '@roots/bud-dashboard'
import {join} from 'path'

const cwd = process.cwd()

/**
 * Describe
 */
export const describe = 'Compile source assets.'

/**
 * Builder
 */
export const builder = ({example}) =>
  example('Build', 'bud build --mode production')
    .hide('help')
    .hide('version')

/**
 * Handler
 */
export const handler = ({config}): void => {
  try {
    require(join(cwd, config ?? 'sage.config.js'))
  } catch (error) {
    Error(error.toString(), `Error`)
  }
}
