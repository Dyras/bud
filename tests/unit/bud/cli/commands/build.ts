import build from '@roots/bud/src/cli/commands/build'

describe('Build Command', () => {
  it('has name: build', () => {
    expect(build.name).toBe('Build')
  })

  it('has description', () => {
    expect(build.description).toBeDefined()
  })

  it(`description is 'compile build assets' `, () => {
    expect(build.description).toBe('compile assets')
  })

  it('has examples array', () => {
    expect(build.examples).toBeInstanceOf(Array)
  })

  it('has expected flags', () => {
    expect(build.flags).toMatchSnapshot({
      cache: {
        allowNo: false,
        char: 'c',
        description: 'cache compiler references to disk',
        parse: expect.any(Function),
        type: 'boolean',
      },
      cli: {
        allowNo: false,
        description: 'non raw mode tty interoperable output',
        parse: expect.any(Function),
        type: 'boolean',
      },
      discover: {
        allowNo: false,
        char: 'd',
        description:
          'automatically utilize installed extensions',
        parse: expect.any(Function),
        type: 'boolean',
      },
      hash: {
        allowNo: false,
        description: 'hash compiled filenames',
        parse: expect.any(Function),
        type: 'boolean',
      },
      help: {
        allowNo: false,
        char: 'h',
        description: 'show CLI help',
        parse: expect.any(Function),
        type: 'boolean',
      },
      install: {
        allowNo: false,
        char: 'i',
        description: 'ensure peer dependencies are installed',
        parse: expect.any(Function),
        type: 'boolean',
      },
      log: {
        allowNo: false,
        char: 'l',
        description: 'log to console',
        parse: expect.any(Function),
        type: 'boolean',
      },
      manifest: {
        allowNo: false,
        description: 'produce a manifest',
        parse: expect.any(Function),
        type: 'boolean',
      },
      minimize: {
        allowNo: false,
        char: 'm',
        description: 'minimize file size of compiled assets',
        parse: expect.any(Function),
        type: 'boolean',
      },
      target: {
        char: 't',
        default: [],
        description: 'limit compilation to this compiler',
        input: [],
        multiple: true,
        parse: expect.any(Function),
        type: 'option',
      },
    })
  })
})
