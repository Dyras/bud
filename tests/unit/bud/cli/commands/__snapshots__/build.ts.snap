// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build Command has expected flags 1`] = `
Object {
  "cache": Object {
    "allowNo": false,
    "char": "c",
    "description": "cache compiler references to disk",
    "parse": Any<Function>,
    "type": "boolean",
  },
  "cli": Object {
    "allowNo": false,
    "description": "non raw mode tty interoperable output",
    "parse": Any<Function>,
    "type": "boolean",
  },
  "discover": Object {
    "allowNo": false,
    "char": "d",
    "description": "automatically utilize installed extensions",
    "parse": Any<Function>,
    "type": "boolean",
  },
  "hash": Object {
    "allowNo": false,
    "description": "hash compiled filenames",
    "parse": Any<Function>,
    "type": "boolean",
  },
  "help": Object {
    "allowNo": false,
    "char": "h",
    "description": "show CLI help",
    "parse": Any<Function>,
    "type": "boolean",
  },
  "install": Object {
    "allowNo": false,
    "char": "i",
    "description": "ensure peer dependencies are installed",
    "parse": Any<Function>,
    "type": "boolean",
  },
  "log": Object {
    "allowNo": false,
    "char": "l",
    "description": "log to console",
    "parse": Any<Function>,
    "type": "boolean",
  },
  "manifest": Object {
    "allowNo": false,
    "description": "produce a manifest",
    "parse": Any<Function>,
    "type": "boolean",
  },
  "minimize": Object {
    "allowNo": false,
    "char": "m",
    "description": "minimize file size of compiled assets",
    "parse": Any<Function>,
    "type": "boolean",
  },
  "target": Object {
    "char": "t",
    "default": Array [],
    "description": "limit compilation to this compiler",
    "input": Array [],
    "multiple": true,
    "parse": Any<Function>,
    "type": "option",
  },
}
`;
